Assets/HoloToolkit/UX/Scripts/Buttons/ButtonIconProfileTexture.cs

```diff
@@ -63,7 +63,7 @@ public class ButtonIconProfileTexture : ButtonIconProfile
         public Texture2D AdjustHologram;
         public Texture2D RemoveHologram;
 
-        [HideInMRDLInspector]
+        [HideInMRTKInspector]
```

Comment:
How come we have special inspector attributes?

Shouldn't the normal `[HideInInspector]` work?

Reply:
This hides fields in MRTKInspectors, but leaves them visible in standard inspectors. This is necessary if you want to display an editable property that gets / sets a serialized field.

Reply:
Why not just write the custom editor inspector needed to handle this?

Not sure what the benefit is here.

Reply:
>Why not just write the custom editor inspector needed to handle this?

@StephenHodgson Darn it Stephen here I am trying to avoid a rabbit hole and you go and ask a direct question like that. I guess we're diving in, lol.

That was our thought, initially - just write custom inspectors. But we found that they're not as dev-friendly as we had assumed. This system is the result of a bunch of testing / debate / feedback about how to do inspectors in the most dev-friendly way.

We found that although custom inspectors make scripts more accessible and guard against user error, they have nasty drawbacks. They're tough to keep homogeneous when they're written by a team, which is bad for comprehension. They often drift out of sync with the scripts they target. They can become _too_ logic-heavy and inter-dependent with their scripts. And they can conceal system logic from developers that would be beneficial to expose.

The extra Editor files are also surprisingly error-prone when devs drag individual component scripts to a new project, sans the framework. You'd think it wouldn't be an issue, but Editor files constantly get left behind, leaving folks to wonder why components look 'different.' It gets worse when Editor classes require still more files - eg they inherit from a base class or use a static Editor utility (a practical necessity when your goal is consistency).

There are other drawbacks too, but those are the biggies.

So this is a stab at eliminating the drawbacks while keeping the benefits. By generating inspectors from attributes we keep them homogeneous (warnings / errors / doc links are visually consistent and appear in the same places no matter who writes them), we discourage logic-heavy inspectors, and what logic does exist is plainly visible right there in the scripts.

A final benefit that dragging single scripts into a new project required only a single additional file - MRTKUtility.cs - instead of all those individual Editor files and base classes etc. etc. This has been sort of nullified by the style requirement of breaking out every class into an individual file, but maybe we can revisit that issue later down the road.

Anyway, that's the idea. It's an ongoing experiment with lots of room for improvement so I'm eager to hear thoughts on how to make it better - BUT, that said, I really, really don't want to distract from the our immediate goal of getting the core functionality merged.

Reply:
>Darn it Stephen here I am trying to avoid a rabbit hole and you go and ask a direct question like that. I guess we're diving in, lol.

@Railboy My bad 🤣 Thanks for the great write up.  I noticed right away the the MRDL inspectors are quite complex.

> The extra Editor files are also surprisingly error-prone when devs drag individual component scripts to a new project.

Whoever these devs are you should really tell them that's a bad idea, unless they're dragging whole folders (and even then it should be organized in such a way that if they drag the scripts folder the dependent editor folder is a subfolder too.  Seriously who just drags over a single script?  Do they edit it down to something else?

Why aren't they using the asset exporter that will grab the dependencies for them?

> They're often too logic-heavy and inter-dependent with their scripts. And they can conceal system logic from developers that would be beneficial to expose.

I can see this being an issue.  You def want to make as much of the logic as transparent as possible.

Reply:
>Do they edit it down to something else? Why not just write it up from scratch then? Why aren't they using the asset exporter that will grab the dependencies for them?

Careful, questions like these will drive you mad, haha. I guess different perspectives? We think of toolkits as monolithic but lots of devs treat them more a la carte, even when big chunks are interdependent.

It's not something I'd recommend, but there's also no reason _in principle_ why Billboard.cs (or whatever) can't be yanked out and used on its own.

So I just say whatever, if enough devs are doing something, and there's no reason in principle why it shouldn't be possible, then our job is to make it less error-prone.

I'd feel differently if these were pure C# toolkits but the whole Unity ecosystem is so 'flexible' that forcing devs not to use it a particular way is like stopping a faucet with your finger.